### ISSUE TITLE
Issue with CDI, Named beans and EAR packaging

### ISSUE THREAD

**Issue by sverkera** [astefanutti/camel-cdi#11]
I have an application deployed on Wildfly 8.2.0 with Wildfly-Camel 2.2.0 and Camel 2.15.0 which is heavily using named beans in camel routes. 

While upgrading the components to Wildfly 9.0.1 and Wildfly-Camel 3.1.0 which uses Camel 2.16.0 I hit the problem that Camel routes are unable to find my named beans. I filed an issue on this to Wildfly-Camel github project: 

https://github.com/wildfly-extras/wildfly-camel/issues/978

and I've filed a question to camel-user:
http://camel.465427.n5.nabble.com/Issue-with-CDI-Named-beans-and-EAR-packaging-td5773897.html

When I found your implementation I tested with that but run into the same issue. Detailed description is availible on the links above.

In short, also your CdiBeanRegistry uses an injected BeanManager which belongs to the camel-cdi jar. In the earlier version of camel-cdi they used Deltaspike to find the BeanManager that belong to the ejb jar which knew about these named beans.

It is not clear to me which is the correct behaviour, if like in the older implementation the BeanManager belonging to ejb jar must be found and queried or if the query shoulde be delegated by appserver from camel-cdi jar's BeanManager and it is that delegation which is failing. 

Which is the correct behaviour and where should the issue be solved?

**Comment by astefanutti** [comment]
I've just answered your question on Camel forum.

To move forward on a potential solution at the level of Camel CDI, would that be possible for you to create a minimal test case that is representative of your EAR structure and that reproduces the issue?

Ideally, that'd be a test in the `envs/ee` module. If your target deployment is WildFly Camel, you may want to adapt the [`CamelWildFlyTest`](https://github.com/astefanutti/camel-cdi/blob/1c6ee996e88c37831c77f1f47aaa4ced858986b6/envs/ee/src/test/java/org/apache/camel/cdi/ee/CamelWildFlyTest.java#L36-L37) class. Or if the target is WildFly and your EAR packages Camel CDI, you may want to adapt the [`CamelEE7Test`](https://github.com/astefanutti/camel-cdi/blob/6719200334a67bc7596363a1c20712944c30b768/envs/ee/src/test/java/org/apache/camel/cdi/ee/CamelEE7Test.java) class.

**Comment by sverkera** [comment]
Right now I did a simple workaround to always look up the BeanManager from jndi in CdiBeanRegistry instead of using the injected bean. That work for my usecase for the time being but of course I want to see a proper solution.

I'll look at creating an ear version of CamelWildflyTest and we'll take it from there.

**Comment by astefanutti** [comment]
Yes, using the bean manager from JNDI is an option.

I'd like to get a better understanding of your test case so that we can decide what's the best strategy to implement in Camel CDI.

**Comment by sverkera** [comment]
The testcases are quite close to the problem I see. The difference of your CamelEE7Test is that there you package all the libs in the ear while I am using the Wildfly modules installed by Wildfly-Camel patch.

If I just changed the packaging I got error that the Deltaspike classes couldn't be found, that is because in org.wildfly.extension.camel.deployment.CamelDependenciesProcessor the Deltaspike modules are added as dependencies if @CamelAware annotation is present.

If I added them as dependencies with jboss-deployment-structure.xml then the timer component could not be found. The way I've found to work is to remove @CamelAware annotation and instead use a jboss-all.xml descriptor to add the same module dependencies except for Deltaspike.
[0002-Added-testcase-for-ear-packaging-using-Camel-from-Wi.txt](https://github.com/astefanutti/camel-cdi/files/39291/0002-Added-testcase-for-ear-packaging-using-Camel-from-Wi.txt)

**Comment by sverkera** [comment]
Attached here is the workaround to look up BeanManager through jndi but since that is not availible for the testcase in camel-cdi I had to fall back to injected bean if jndi is not availible.

Deltaspike's BeanManagerProvider and Solder's BeanManagerLocator does similar trick, but covering more cases.

**Comment by astefanutti** [comment]
After applying both your patches above, compiling and running `mvn test -pl envs/ee -Pwildfly-camel -Dtest=CamelWildFlyEarTest`, I still faced the following exception:
```
org.apache.camel.ResolveEndpointFailedException: Failed to resolve endpoint: timer://timer1?period=1000 due to: No component found with scheme: timer
```
And had to add the `@CamelAware` annotation to have it working.

I've pushed the [`CamelWildFlyEarTest`](https://github.com/astefanutti/camel-cdi/blob/9487bc4548e84f55ed01aa0db634bf799c46fe23/envs/ee/src/test/java/org/apache/camel/cdi/ee/CamelWildFlyEarTest.java) test in 9487bc4548e84f55ed01aa0db634bf799c46fe23, and the use of the bean manager from JNDI in 584b37b65bb11796e03bf2170a1838e970996489. The test can be run with `mvn test -pl envs/ee -Pwildfly-camel -Dtest=CamelWildFlyEarTest`.

My current understanding is that the issue is two fold:

- The [`timer`](https://github.com/astefanutti/camel-cdi/blob/9487bc4548e84f55ed01aa0db634bf799c46fe23/envs/ee/src/main/java/org/apache/camel/cdi/ee/Bootstrap.java#L52) component is not a bean and thus not retrieved via the CDI Camel registry. It is looked up via the standard Camel `ComponentResolver` mechanism which _searches_ for the `META-INF/services/org/apache/camel/component/timer` resource using the `ClassResolver` abstraction. When running the test with WildFly Camel, the `ClassResolver`, which is an instance of `org.wildly.extension.camel.handler.WildFlyClassResolver`, does not return any resource when called on `classResolver.loadResourceAsStream(uri)` with `uri` being ``META-INF/services/org/apache/camel/component/timer` and that belongs to `camel-core`. That is fixed (and supposed to be I'd said) by adding the `@CamelAware` annotation. Then the DeltaSpike `NoClassDefFoundError` happens and is fixed by adding the `jboss-all.xml` descriptor. It looks like the `jboss-deployment-structure.xml` descriptor is not necessary. So there is still the open question to be discussed in relation to wildfly-extras/wildfly-camel#978 why the `jboss-all.xml` descriptor is necessary.
- The [`helloCamel`](https://github.com/astefanutti/camel-cdi/blob/9487bc4548e84f55ed01aa0db634bf799c46fe23/envs/ee/src/main/java/org/apache/camel/cdi/ee/HelloCamel.java#L27) bean is retrieved via the CDI Camel registry. When the test is executed in WildFly without WildFly Camel, the injected bean manager below, which is the one that Weld has assigned for the extension in the `camel-cdi.jar` library, is capable of getting an instance of the bean:

    ```
    Weld BeanManager for camel-ee7.ear/lib/camel-cdi-1.2-SNAPSHOT.jar [bean count=40]
    ```
    While when the test is executed with WildFly Camel, the same bean manager below is not capable of retrieving an instance of the bean:

    ```
    Weld BeanManager for camel-wildfly.ear.external.jar:file:/Users/astefanu/Development/camel-cdi/envs/ee/target/wildfly-camel/wildfly-9.0.1.Final/modules/system/layers/fuse/org/apache/camel/component/cdi/main/camel-cdi-1.2-SNAPSHOT.jar!/META-INF/beans.xml [bean count=40]
    ```
    And the bean manager from JNDI below, which is the one Weld has assigned to the application module, is capable of getting an instance of the bean:

    ```
    Weld BeanManager for camel-wildfly.ear/camel-wildfly.jar/ [bean count=32]
    ```
    So, there is clearly a behavioural difference that's introduced by WildFly Camel though from the [Obtaining a `Bean` by name](http://docs.jboss.org/cdi/spec/1.2/cdi-spec.html#bm_obtain_bean_by_name) section of the CDI specification, I'd said the WildFly Camel behaviour is actually the correct one. I'll continue investigating on it.

**Comment by sverkera** [comment]
On your first item, I got it working with either @CamelAware + jboss-all.xml OR jboss-all.xml + jboss-deployment-structure.xml. I agree that the former should be the more correct, but tracing it seems that having the camel settings jboss-all.xml makes CamelDependenciesProcessor not process that annotation but it may affect something somewhere else.

**Comment by astefanutti** [comment]
I suggest we iterate on first item with @jamesnetherton as I understand a Camel CDI application should not deal with those descriptors when deployed in WildFly Camel.

**Comment by sverkera** [comment]
Agree on that

**Comment by jamesnetherton** [comment]
`jboss-all.xml` is only required if you want to customise which camel components are enabled for your application. Otherwise, detection of `@CamelAware` or `@ContextName` annotations is the trigger for wildfly-camel to add camel dependencies to deployments.  

I tested a modified version of `CamelWildFlyEarTest` with the `jboss-all` and `jboss-deployment-structure` descriptors commented out. As expected I hit some NCDFE issues

```
Caused by: java.lang.ClassNotFoundException: org.apache.camel.model.ModelCamelContext from [Module \"deployment.camel-wildfly.ear:main\" from Service Module Loader]
```

Notice the module mentioned in the exception is the Ear deployment and not the Jar deployment where the `@CamelAware` annotation is applied. So I put some extra logging into the subsystem processors:

```
15:01:15,275 WARN  [org.wildfly.extension.camel] (MSC service thread 1-3) >>>>>>>>>>> CAMEL NOT ENABLED camel-wildfly.ear
15:01:15,276 WARN  [org.wildfly.extension.camel] (MSC service thread 1-3) >>>>>>>>>>> CAMEL NOT ENABLED test.war
15:01:15,276 WARN  [org.wildfly.extension.camel] (MSC service thread 1-7) @CamelAware annotation found camel-wildfly.jar
15:01:15,277 WARN  [org.wildfly.extension.camel] (MSC service thread 1-7) >>>>>>>>>>> CAMEL ENABLED camel-wildfly.jar
```

This shows that the Camel dependencies are added to the Jar deployment but not to the Ear deployment. So I hacked some code into the `CamelDependenciesProcessor` to add the Camel dependencies to the parent deployment (the Ear). When I run the test it seems to complete successfully.

```
15:11:36,434 INFO  [org.apache.camel.cdi.ee.Bootstrap] (ServerService Thread Pool -- 23) >> Camel context created and Camel route started.
15:11:37,226 INFO  [org.jboss.as.server] (management-handler-thread - 3) WFLYSRV0010: Deployed "camel-wildfly.ear" (runtime-name : "camel-wildfly.ear")
15:11:37,447 INFO  [route1] (Camel (camel-cdi) thread #0 - timer://timer1) >> Response : >> Hello Camel user.
15:11:38,434 INFO  [route1] (Camel (camel-cdi) thread #0 - timer://timer1) >> Response : >> Hello Camel user.
15:11:39,434 INFO  [route1] (Camel (camel-cdi) thread #0 - timer://timer1) >> Response : >> Hello Camel user.
```

I guess this points to a problem with how parent / child deployments are handled in WildFly-Camel.

**Comment by astefanutti** [comment]
Based on the latest analysis done by @jamesnetherton summed up in https://github.com/wildfly-extras/wildfly-camel/issues/978#issuecomment-158971039, this issue is to be further followed up in wildfly-extras/wildfly-camel#978.

**Comment by astefanutti** [comment]
This is resolved with wildfly-extras/wildfly-camel@f68932bfdecc8efc1392882ecc187ca97af854f6.

**Comment by ke5stl** [comment]
it appears that this issue exists also for glassfish 4.x.x

please advise as to what the solution is!

**Comment by astefanutti** [comment]
@ke5stl the final outcome of that particular issue has been a fix at the level of [WildFly Camel](https://github.com/wildfly-extras/wildfly-camel).

Are you using that particular version of Camel CDI or the component from the Camel project? If yes, would you mind creating a new issue so that we can better iterate on the GlassFish support?

**Comment by ke5stl** [comment]
I am solely using the Camel CDI, without any WildFly specific add-ons, as I'm using Glassfish, and I have no expectation that WildFly specific things will work for Glassfish.  I suppose this means I need to open a ticket for Glassfish.  I will do that.

**Comment by ke5stl** [comment]
I opened a ticket for the Glassfish centric but same issue.  Do I need to do more, to get some progress on it?  Please advise.

Best Regards

**Comment by astefanutti** [comment]
@ke5stl I've just answered you directly in #13.
